### ISSUE TITLE
Implementation with Firebase

### ISSUE THREAD

**Issue by rugor** [alexsasharegan/vue-transmit#26]
Hi, vue-transmit looks great. I want to use it in a project that uses Firebase, but Firebase accepts uploads without a url endpoint. When I leave the url prop empty or leave it out altogether I get errors. Do you have a suggestion on how to get around this? Maybe its easy but I haven't figured it out yet.

Firebase accepts uploads [this way](https://firebase.google.com/docs/storage/web/upload-files#full_example) (example from Firebase docs) which is how I am doing the upload.

**Comment by alexsasharegan** [comment]
A really quick overview of the docs you provided make me think I could possibly set up some kind of adapter for the actual upload process. Then you could any SDK, but you would be responsible for firing the lifecycle hooks during upload (progress, completion, error, etc.). I'll try to take a look at this tonight.

I think currently there is a way to do this by turning off `autoProcessQueue`. Here is a rough sketch

```js
{
	data() {
		return {
			isProcessing: false,
		}
	},
	watch: {
		queuedFiles(newFiles) {
			if (newFiles.length > 0) {
				this.processFiles()
			}
		},
	},
	computed: {
		queuedFiles() {
			// assumes you put the ref "vtransmit" on your instance
			return this.$refs.vtransmit.queuedFiles
		},
	},
	methods: {
		processFiles(vtFiles) {
			if (this.isProcessing) {
				return
			}

			this.isProcessing = true
			let $vt = this.$refs.vtransmit
			let vtFiles = this.queuedFiles // cache this list since it can update during upload
			for (let vtFile of vtFiles) {
				vtFile.status = "uploading"
				let fbFile = vtFile.nativeFile
				let metadata = {
					contentType: vtFile.type,
				}
				let uploadTask = storageRef.child("images/" + file.name).put(file, metadata)
				// Listen for state changes, errors, and completion of the upload.
				uploadTask.on(
					firebase.storage.TaskEvent.STATE_CHANGED, // or 'state_changed'
					function(snapshot) {
						// Get task progress, including the number of bytes uploaded and the total number of bytes to be uploaded
						var progress = snapshot.bytesTransferred / snapshot.totalBytes * 100

						// If the "snapshot" is a progress event, the vtFile can handle it
						vtFile.handleProgress(snapshot)
						$vt.$emit("upload-progress", vtFile, progress, snapshot.bytesTransferred)

						console.log("Upload is " + progress + "% done")
						switch (snapshot.state) {
							case firebase.storage.TaskState.PAUSED: // or 'paused'
								console.log("Upload is paused")
								break
							case firebase.storage.TaskState.RUNNING: // or 'running'
								console.log("Upload is running")
								break
						}
					},
					function(error) {
						// not sure what the xhr would be here
						$vt.errorProcessing(vtFiles, "some message", xhr)
						// A full list of error codes is available at
						// https://firebase.google.com/docs/storage/web/handle-errors
						switch (error.code) {
							case "storage/unauthorized":
								// User doesn't have permission to access the object
								break

							case "storage/canceled":
								// User canceled the upload
								break

							// ...

							case "storage/unknown":
								// Unknown error occurred, inspect error.serverResponse
								break
						}
					},
					function() {
						vtFile.status = "success"
						// Upload completed successfully, now we can get the download URL
						var downloadURL = uploadTask.snapshot.downloadURL
					}
				)
			}

			this.isProcessing = false
		},
	},
}
```

**Comment by rugor** [comment]
Thanks for the example and quick follow-up. I'll keep on with what you provided and see if I can get it going. As for the interceptor as a way to interact with sdk's I think that would be really useful.

**Comment by alexsasharegan** [comment]
I really like the idea of making the uploader an adapter plugin here. As I'm digging in, I think it may take some time to get right, heads up.

I plan on pulling the current XHR uploader out into an adapter to pave the way. This should force me into making reasonable API choices. Once I've got a working version, would be interested in contributing a Firebase adapter? With ESM, this could be an easy import only tool, not sacrificing library payload size for the main use case.

Also, feel free to drop questions as you tackle the workaround path.

**Comment by alexsasharegan** [comment]
Is the error you're talking about the url required prop? Go ahead and put a dummy url in there to keep the vue error quiet. You don't have to do anything with the XHR in the error handling. It just gets emitted in the error event. Since you're taking control of all this, you have access to the guts in your component code anyway.

This PR is challenging because it really requires some rethinking of the component options as props. Feel free to look it over, but I've got a lot of work to do still before it's done. I would also like to get testing going for the inner mechanics. It should make it easier for any plugins to verify their interoperability.

**Comment by alexsasharegan** [comment]
Any updates @rugor? I've still got a ways to go on the new version, but I've got the interface almost down to just a class with two methods: `cancelUpload` & `uploadFiles`. The upload func just returns a promise, or rejects with an object containing the event to fire and a message. Any upload transport adapter can have its own options bag that vue-transmit will pass to the adapter's constructor. I've pushed the XHR adapter up to the PR.
